README.md
# BIP39-WEB

[![NPM](https://img.shields.io/npm/v/bip39-web.svg)](https://www.npmjs.org/package/bip39-web)

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


JavaScript implementation of [Bitcoin BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki): Mnemonic code for generating deterministic keys

## Reminder for developers

***Please remember to allow recovery from mnemonic phrases that have invalid checksums (or that you don't have the wordlist)***

When a checksum is invalid, warn the user that the phrase is not something generated by your app, and ask if they would like to use it anyway. This way, your app only needs to hold the wordlists for your supported languages, but you can recover phrases made by other apps in other languages.

However, there should be other checks in place, such as checking to make sure the user is inputting 12 words or more separated by a space. ie. `phrase.trim().split(/\s+/g).length >= 12`

## Removing wordlists from webpack/browserify

 Browserify/Webpack bundles can get very large if you include all the wordlists, so you can now exclude wordlists to make your bundle lighter.

 For example, if we want to exclude all wordlists besides chinese_simplified, you could build using the browserify command below.

 ```bash
$ browserify -r bip39-web -s bip39-web \
  --exclude=./wordlists/english.json \
  --exclude=./wordlists/japanese.json \
  --exclude=./wordlists/spanish.json \
  --exclude=./wordlists/italian.json \
  --exclude=./wordlists/french.json \
  --exclude=./wordlists/korean.json \
  --exclude=./wordlists/czech.json \
  --exclude=./wordlists/portuguese.json \
  --exclude=./wordlists/chinese_traditional.json \
   > bip39-web.browser.js
```

 This will create a bundle that only contains the chinese_simplified wordlist, and it will be the default wordlist for all calls without explicit wordlists.

 You can also do this in Webpack using the `IgnorePlugin`. Here is an example of excluding all non-English wordlists

 ```javascript
 ...
 plugins: [
   new webpack.IgnorePlugin(/^\.\/wordlists\/(?!english)/, /bip39\/src$/),
 ],
 ...
 ```

 This is how it will look in the browser console.

 ```javascript
> bip39.entropyToMnemonic('00000000000000000000000000000000')
"的 的 的 的 的 的 的 的 的 的 的 在"
> bip39.wordlists.chinese_simplified
Array(2048) [ "的", "一", "是", "在", "不", "了", "有", "和", "人", "这", … ]
> bip39.wordlists.english
undefined
> bip39.wordlists.japanese
undefined
> bip39.wordlists.spanish
undefined
> bip39.wordlists.italian
undefined
> bip39.wordlists.french
undefined
> bip39.wordlists.korean
undefined
> bip39.wordlists.czech
undefined
> bip39.wordlists.portuguese
undefined
> bip39.wordlists.chinese_traditional
undefined
```

 For a list of supported wordlists check the wordlists folder. The name of the json file (minus the extension) is the name of the key to access the wordlist.

 You can also change the default wordlist at runtime if you dislike the wordlist you were given as default.

 ```javascript
> await bip39.entropyToMnemonicAsync('00000000000000000000000000000fff')
"あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あまい　ろんり"
> bip39.setDefaultWordlist('italian')
undefined
> await bip39.entropyToMnemonicAsync('00000000000000000000000000000fff')
"abaco abaco abaco abaco abaco abaco abaco abaco abaco abaco aforisma zibetto"
```

## Installation
``` bash
npm install bip39-web
```

## Examples
``` js
// Generate a random mnemonic (uses crypto.randomBytes under the hood), defaults to 128-bits of entropy
const mnemonic = await bip39.generateMnemonicAsync()
// => 'seed sock milk update focus rotate barely fade car face mechanic mercy'

(await bip39.mnemonicToSeedAsync('basket actual')).toString('hex')
// => '5cf2d4a8b0355e90295bdfc565a022a409af063d5365bb57bf74d9528f494bfa4400f53d8349b80fdae44082d7f9541e1dba2b003bcfec9d0d53781ca676651f'

await bip39.mnemonicToSeedAsync('basket actual')
// => <Buffer 5c f2 d4 a8 b0 35 5e 90 29 5b df c5 65 a0 22 a4 09 af 06 3d 53 65 bb 57 bf 74 d9 52 8f 49 4b fa 44 00 f5 3d 83 49 b8 0f da e4 40 82 d7 f9 54 1e 1d ba 2b ...>

// mnemonicToSeed has an synchronous version
// mnemonicToSeedSync is less performance oriented
bip39.mnemonicToSeedAsync('basket actual').then(console.log)
// => <Buffer 5c f2 d4 a8 b0 35 5e 90 29 5b df c5 65 a0 22 a4 09 af 06 3d 53 65 bb 57 bf 74 d9 52 8f 49 4b fa 44 00 f5 3d 83 49 b8 0f da e4 40 82 d7 f9 54 1e 1d ba 2b ...>

bip39.mnemonicToSeedAsync('basket actual').then(bytes => bytes.toString('hex')).then(console.log)
// => '5cf2d4a8b0355e90295bdfc565a022a409af063d5365bb57bf74d9528f494bfa4400f53d8349b80fdae44082d7f9541e1dba2b003bcfec9d0d53781ca676651f'

await bip39.mnemonicToSeedAsync('basket actual', 'a password')
// => <Buffer 46 16 a4 4f 2c 90 b9 69 02 14 b8 fd 43 5b b4 14 62 43 de 10 7b 30 87 59 0a 3b b8 d3 1b 2f 3a ef ab 1d 4b 52 6d 21 e5 0a 04 02 3d 7a d0 66 43 ea 68 3b ... >

await bip39.validateMnemonicAsync(mnemonic)
// => true

await bip39.validateMnemonicAsync('basket actual')
// => false
```


``` js
const bip39 = require('bip39-web')

// defaults to BIP39 English word list
// uses HEX strings for entropy
const mnemonic = await bip39.entropyToMnemonicAsync('00000000000000000000000000000000')
// => abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about

// reversible
await bip39.mnemonicToEntropyAsync(mnemonic)
// => '00000000000000000000000000000000'
```


CHANGELOG.md
# 1.0.0
__breaking changes__
- Ported original bip39 package to support running within web browsers.
- All methods are now Async

# 3.0.0
__added__
- Added TypeScript support (#104)
- Added support for excluding wordlists from packages (#105)

__changed__
- Changed `mnemonicToSeed` to use async, sync version moved to `mnemonicToSeedSync` (#104)

__removed__
- Removed explicit hex methods (use `toString('hex')` on the Buffer) (#104)

# 2.3.1

__breaking changes__

9-letter mnemonics can no longer be geerated and calling `validateMnemonic` will always return false. This was [fixed in the BIP as a patch](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic), so we had to follow suite.


test/index.js
var bip39 = require('../')
var download = require('../util/wordlists').download
var WORDLISTS = {
  english: require('../src/wordlists/english.json'),
  japanese: require('../src/wordlists/japanese.json'),
  custom: require('./wordlist.json')
}

var vectors = require('./vectors.json')
var test = require('tape')

function testVector (description, wordlist, password, v, i) {
  var ventropy = v[0]
  var vmnemonic = v[1]
  var vseedHex = v[2]

  test('for ' + description + '(' + i + '), ' + ventropy, function (t) {
    t.plan(5)

    const fn = async () => {
      t.equal(await bip39.mnemonicToEntropyAsync(vmnemonic, wordlist), ventropy, 'mnemonicToEntropy returns ' + ventropy.slice(0, 40) + '...')
      t.equal((await bip39.mnemonicToSeedAsync(vmnemonic, password)).toString('hex'), vseedHex, 'mnemonicToSeedAsync returns ' + vseedHex.slice(0, 40) + '...')
      t.equal(await bip39.entropyToMnemonicAsync(ventropy, wordlist), vmnemonic, 'entropyToMnemonicAsync returns ' + vmnemonic.slice(0, 40) + '...')

      function rng () { return Buffer.from(ventropy, 'hex') }
      t.equal(await bip39.generateMnemonicAsync(undefined, rng, wordlist), vmnemonic, 'generateMnemonicAsync returns RNG entropy unmodified')
      t.equal(await bip39.validateMnemonicAsync(vmnemonic, wordlist), true, 'validateMnemonicAsync returns true')
    };
    fn();
  })
}

vectors.english.forEach(function (v, i) { testVector('English', undefined, 'TREZOR', v, i) })
vectors.japanese.forEach(function (v, i) { testVector('Japanese', WORDLISTS.japanese, '㍍ガバヴァぱばぐゞちぢ十人十色', v, i) })
vectors.custom.forEach(function (v, i) { testVector('Custom', WORDLISTS.custom, undefined, v, i) })

test('getDefaultWordlist returns "english"', function (t) {
  t.plan(1)
  const english = bip39.getDefaultWordlist()
  t.equal(english, 'english')
  // TODO: Test that Error throws when called if no wordlists are compiled with bip39
})

test('setDefaultWordlist changes default wordlist', async function (t) {
  t.plan(4)
  const english = bip39.getDefaultWordlist()
  t.equal(english, 'english')

  bip39.setDefaultWordlist('italian')

  const italian = bip39.getDefaultWordlist()
  t.equal(italian, 'italian')

  const phraseItalian = await bip39.entropyToMnemonicAsync('00000000000000000000000000000000')
  t.equal(phraseItalian.slice(0, 5), 'abaco')

  bip39.setDefaultWordlist('english')

  const phraseEnglish = await bip39.entropyToMnemonicAsync('00000000000000000000000000000000')
  t.equal(phraseEnglish.slice(0, 7), 'abandon')
})

test('setDefaultWordlist throws on unknown wordlist', function (t) {
  t.plan(2)
  const english = bip39.getDefaultWordlist()
  t.equal(english, 'english')

  try {
    bip39.setDefaultWordlist('abcdefghijklmnop')
  } catch (error) {
    t.equal(error.message, 'Could not find wordlist for language "abcdefghijklmnop"')
    return
  }
  t.assert(false)
})

test('invalid entropy', async function (t) {
  t.plan(3)

  try {
      await bip39.entropyToMnemonicAsync(Buffer.from('', 'hex'));
      t.fail('throws for empty entropy');
  } catch (err) {
    t.equal(err.message, 'Invalid entropy', 'throws for empty entropy');
  }

  try {
    await bip39.entropyToMnemonicAsync(Buffer.from('000000', 'hex'));
    t.fail('throws for entropy that\'s not a multitude of 4 bytes');
  } catch (err) {
    t.equal(err.message, 'Invalid entropy', 'throws for entropy that\'s not a multitude of 4 bytes');
  }

  try {
    await bip39.entropyToMnemonicAsync(Buffer.from(new Array(1028 + 1).join('00'), 'hex'));
    t.fail('throws for entropy that is larger than 1024');
  } catch (err) {
    t.equal(err.message, 'Invalid entropy', 'throws for entropy that is larger than 1024');
  }
})

test('UTF8 passwords', function (t) {
  t.plan(vectors.japanese.length * 2)

  vectors.japanese.forEach(async function (v) {
    var vmnemonic = v[1]
    var vseedHex = v[2]

    var password = '㍍ガバヴァぱばぐゞちぢ十人十色'
    var normalizedPassword = 'メートルガバヴァぱばぐゞちぢ十人十色'

    t.equal((await bip39.mnemonicToSeedAsync(vmnemonic, password)).toString('hex'), vseedHex, 'mnemonicToSeedSync normalizes passwords')
    t.equal((await bip39.mnemonicToSeedAsync(vmnemonic, normalizedPassword)).toString('hex'), vseedHex, 'mnemonicToSeedSync leaves normalizes passwords as-is')
  })
})

test('generateMnemonic can vary entropy length', async function (t) {
  var words = (await bip39.generateMnemonicAsync(160)).split(' ')

  t.plan(1)
  t.equal(words.length, 15, 'can vary generated entropy bit length')
})

test('generateMnemonic requests the exact amount of data from an RNG', async function (t) {
  t.plan(1)

  await bip39.generateMnemonicAsync(160, function (size) {
    t.equal(size, 160 / 8)
    return Buffer.allocUnsafe(size)
  })
})

test('validateMnemonic', async function (t) {
  t.plan(5)

  t.equal(await bip39.validateMnemonicAsync('sleep kitten'), false, 'fails for a mnemonic that is too short')
  t.equal(await bip39.validateMnemonicAsync('sleep kitten sleep kitten sleep kitten'), false, 'fails for a mnemonic that is too short')
  t.equal(await bip39.validateMnemonicAsync('abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about end grace oxygen maze bright face loan ticket trial leg cruel lizard bread worry reject journey perfect chef section caught neither install industry'), false, 'fails for a mnemonic that is too long')
  t.equal(await bip39.validateMnemonicAsync('turtle front uncle idea crush write shrug there lottery flower risky shell'), false, 'fails if mnemonic words are not in the word list')
  t.equal(await bip39.validateMnemonicAsync('sleep kitten sleep kitten sleep kitten sleep kitten sleep kitten sleep kitten'), false, 'fails for invalid checksum')
})

test('exposes standard wordlists', function (t) {
  t.plan(2)
  t.same(bip39.wordlists.EN, WORDLISTS.english)
  t.equal(bip39.wordlists.EN.length, 2048)
})

test('verify wordlists from https://github.com/bitcoin/bips/blob/master/bip-0039/bip-0039-wordlists.md', function (t) {
  download().then(function (wordlists) {
    Object.keys(wordlists).forEach(function (name) {
      t.same(bip39.wordlists[name], wordlists[name])
    })

    t.end()
  })
})


test/readme.js
const bip39 = require('../')
const Buffer = require('safe-buffer').Buffer
const proxyquire = require('proxyquire')
const test = require('tape')

test('README example 1', async function (t) {
  // defaults to BIP39 English word list
  const entropy = 'ffffffffffffffffffffffffffffffff'
  const mnemonic = await bip39.entropyToMnemonicAsync(entropy)

  t.plan(2)
  t.equal(mnemonic, 'zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo wrong')

  // reversible
  t.equal(await bip39.mnemonicToEntropyAsync(mnemonic), entropy)
})

test('README example 2', async function (t) {
  const mnemonic = 'basket actual'
  const seed = await bip39.mnemonicToSeedAsync(mnemonic)

  t.plan(2)
  t.equal(seed.toString('hex'), '5cf2d4a8b0355e90295bdfc565a022a409af063d5365bb57bf74d9528f494bfa4400f53d8349b80fdae44082d7f9541e1dba2b003bcfec9d0d53781ca676651f')
  t.equal(await bip39.validateMnemonicAsync(mnemonic), false)
})


ts_src/index.ts
import { _default as _DEFAULT_WORDLIST, wordlists } from './_wordlists';

let DEFAULT_WORDLIST: string[] | undefined = _DEFAULT_WORDLIST;

const INVALID_MNEMONIC = 'Invalid mnemonic';
const INVALID_ENTROPY = 'Invalid entropy';
const INVALID_CHECKSUM = 'Invalid mnemonic checksum';
const WORDLIST_REQUIRED =
  'A wordlist is required but a default could not be found.\n' +
  'Please pass a 2048 word array explicitly.';

const webcrypto =
  typeof crypto !== 'undefined' ? crypto : require('crypto').webcrypto;

function normalize(str?: string): string {
  return (str || '').normalize('NFKD');
}

function lpad(str: string, padString: string, length: number): string {
  while (str.length < length) {
    str = padString + str;
  }
  return str;
}

function binaryToByte(bin: string): number {
  return parseInt(bin, 2);
}

function bytesToBinary(bytes: number[]): string {
  return bytes.map((x: number): string => lpad(x.toString(2), '0', 8)).join('');
}

async function deriveChecksumBitsAsync(entropyBuffer: Buffer): Promise<string> {
  const ENT = entropyBuffer.length * 8;
  const CS = ENT / 32;
  const hash = await webcrypto.subtle.digest(
    'SHA-256',
    new Uint8Array(
      entropyBuffer.buffer,
      entropyBuffer.byteOffset,
      entropyBuffer.byteLength,
    ),
  );

  return bytesToBinary(Array.from(new Uint8Array(hash))).slice(0, CS);
}

function salt(password?: string): string {
  return 'mnemonic' + (password || '');
}

export async function mnemonicToSeedAsync(
  mnemonic: string,
  password?: string,
): Promise<Buffer> {
  const mnemonicBuffer = Buffer.from(normalize(mnemonic), 'utf8');
  const saltBuffer = Buffer.from(salt(normalize(password)), 'utf8');

  const key = await webcrypto.subtle.importKey(
    'raw',
    new Uint8Array(
      mnemonicBuffer.buffer,
      mnemonicBuffer.byteOffset,
      mnemonicBuffer.byteLength,
    ),
    'PBKDF2', // { name: 'PBKDF2' },
    false,
    ['deriveBits', 'deriveKey'],
  );

  const seedBits = await webcrypto.subtle.deriveBits(
    {
      name: 'PBKDF2',
      salt: new Uint8Array(
        saltBuffer.buffer,
        saltBuffer.byteOffset,
        saltBuffer.byteLength,
      ),
      iterations: 2048,
      hash: 'SHA-512',
    },
    key,
    512,
  );

  return Buffer.from(seedBits);
}

export async function mnemonicToEntropyAsync(
  mnemonic: string,
  wordlist?: string[],
): Promise<string> {
  wordlist = wordlist || DEFAULT_WORDLIST;
  if (!wordlist) {
    throw new Error(WORDLIST_REQUIRED);
  }

  const words = normalize(mnemonic).split(' ');
  if (words.length % 3 !== 0) {
    throw new Error(INVALID_MNEMONIC);
  }

  // convert word indices to 11 bit binary strings
  const bits = words
    .map(
      (word: string): string => {
        const index = wordlist!.indexOf(word);
        if (index === -1) {
          throw new Error(INVALID_MNEMONIC);
        }

        return lpad(index.toString(2), '0', 11);
      },
    )
    .join('');

  // split the binary string into ENT/CS
  const dividerIndex = Math.floor(bits.length / 33) * 32;
  const entropyBits = bits.slice(0, dividerIndex);
  const checksumBits = bits.slice(dividerIndex);

  // calculate the checksum and compare
  const entropyBytes = entropyBits.match(/(.{1,8})/g)!.map(binaryToByte);
  if (entropyBytes.length < 16) {
    throw new Error(INVALID_ENTROPY);
  }
  if (entropyBytes.length > 32) {
    throw new Error(INVALID_ENTROPY);
  }
  if (entropyBytes.length % 4 !== 0) {
    throw new Error(INVALID_ENTROPY);
  }

  const entropy = Buffer.from(entropyBytes);
  const newChecksum = await deriveChecksumBitsAsync(entropy);
  if (newChecksum !== checksumBits) {
    console.log(entropy.toString('hex'));

    console.log('Expected ' + checksumBits + ' actual: ' + newChecksum);
    throw new Error(INVALID_CHECKSUM);
  }

  return entropy.toString('hex');
}

export async function entropyToMnemonicAsync(
  entropy: Buffer | string,
  wordlist?: string[],
): Promise<string> {
  if (!Buffer.isBuffer(entropy)) {
    entropy = Buffer.from(entropy, 'hex');
  }
  wordlist = wordlist || DEFAULT_WORDLIST;
  if (!wordlist) {
    throw new Error(WORDLIST_REQUIRED);
  }

  // 128 <= ENT <= 256
  if (entropy.length < 16) {
    throw new TypeError(INVALID_ENTROPY);
  }
  if (entropy.length > 32) {
    throw new TypeError(INVALID_ENTROPY);
  }
  if (entropy.length % 4 !== 0) {
    throw new TypeError(INVALID_ENTROPY);
  }

  const entropyBits = bytesToBinary(Array.from(entropy));
  const checksumBits = await deriveChecksumBitsAsync(entropy);

  const bits = entropyBits + checksumBits;
  const chunks = bits.match(/(.{1,11})/g)!;
  const words = chunks.map(
    (binary: string): string => {
      const index = binaryToByte(binary);
      return wordlist![index];
    },
  );

  return wordlist[0] === '\u3042\u3044\u3053\u304f\u3057\u3093' // Japanese wordlist
    ? words.join('\u3000')
    : words.join(' ');
}

export async function generateMnemonicAsync(
  strength?: number,
  rng?: (size: number) => Buffer,
  wordlist?: string[],
): Promise<string> {
  strength = strength || 128;
  if (strength % 32 !== 0) {
    throw new TypeError(INVALID_ENTROPY);
  }

  if (!rng) {
    rng = (size: number): Buffer => {
      const rngValues = new Uint8Array(size);
      webcrypto.getRandomValues(rngValues);
      return Buffer.from(rngValues);
    };
  }

  return entropyToMnemonicAsync(rng(strength / 8), wordlist);
}

export async function validateMnemonicAsync(
  mnemonic: string,
  wordlist?: string[],
): Promise<boolean> {
  try {
    await mnemonicToEntropyAsync(mnemonic, wordlist);
  } catch (e) {
    return false;
  }

  return true;
}

export function setDefaultWordlist(language: string): void {
  const result = wordlists[language];
  if (result) {
    DEFAULT_WORDLIST = result;
  } else {
    throw new Error('Could not find wordlist for language "' + language + '"');
  }
}

export function getDefaultWordlist(): string {
  if (!DEFAULT_WORDLIST) {
    throw new Error('No Default Wordlist set');
  }
  return Object.keys(wordlists).filter(
    (lang: string): boolean => {
      if (lang === 'JA' || lang === 'EN') {
        return false;
      }
      return wordlists[lang].every(
        (word: string, index: number): boolean =>
          word === DEFAULT_WORDLIST![index],
      );
    },
  )[0];
}

export { wordlists } from './_wordlists';


src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _wordlists_1 = require("./_wordlists");
let DEFAULT_WORDLIST = _wordlists_1._default;
const INVALID_MNEMONIC = 'Invalid mnemonic';
const INVALID_ENTROPY = 'Invalid entropy';
const INVALID_CHECKSUM = 'Invalid mnemonic checksum';
const WORDLIST_REQUIRED = 'A wordlist is required but a default could not be found.\n' +
    'Please pass a 2048 word array explicitly.';
const webcrypto = typeof crypto !== 'undefined' ? crypto : require('crypto').webcrypto;
function normalize(str) {
    return (str || '').normalize('NFKD');
}
function lpad(str, padString, length) {
    while (str.length < length) {
        str = padString + str;
    }
    return str;
}
function binaryToByte(bin) {
    return parseInt(bin, 2);
}
function bytesToBinary(bytes) {
    return bytes.map((x) => lpad(x.toString(2), '0', 8)).join('');
}
async function deriveChecksumBitsAsync(entropyBuffer) {
    const ENT = entropyBuffer.length * 8;
    const CS = ENT / 32;
    const hash = await webcrypto.subtle.digest('SHA-256', new Uint8Array(entropyBuffer.buffer, entropyBuffer.byteOffset, entropyBuffer.byteLength));
    return bytesToBinary(Array.from(new Uint8Array(hash))).slice(0, CS);
}
function salt(password) {
    return 'mnemonic' + (password || '');
}
async function mnemonicToSeedAsync(mnemonic, password) {
    const mnemonicBuffer = Buffer.from(normalize(mnemonic), 'utf8');
    const saltBuffer = Buffer.from(salt(normalize(password)), 'utf8');
    const key = await webcrypto.subtle.importKey('raw', new Uint8Array(mnemonicBuffer.buffer, mnemonicBuffer.byteOffset, mnemonicBuffer.byteLength), 'PBKDF2', // { name: 'PBKDF2' },
    false, ['deriveBits', 'deriveKey']);
    const seedBits = await webcrypto.subtle.deriveBits({
        name: 'PBKDF2',
        salt: new Uint8Array(saltBuffer.buffer, saltBuffer.byteOffset, saltBuffer.byteLength),
        iterations: 2048,
        hash: 'SHA-512',
    }, key, 512);
    return Buffer.from(seedBits);
}
exports.mnemonicToSeedAsync = mnemonicToSeedAsync;
async function mnemonicToEntropyAsync(mnemonic, wordlist) {
    wordlist = wordlist || DEFAULT_WORDLIST;
    if (!wordlist) {
        throw new Error(WORDLIST_REQUIRED);
    }
    const words = normalize(mnemonic).split(' ');
    if (words.length % 3 !== 0) {
        throw new Error(INVALID_MNEMONIC);
    }
    // convert word indices to 11 bit binary strings
    const bits = words
        .map((word) => {
        const index = wordlist.indexOf(word);
        if (index === -1) {
            throw new Error(INVALID_MNEMONIC);
        }
        return lpad(index.toString(2), '0', 11);
    })
        .join('');
    // split the binary string into ENT/CS
    const dividerIndex = Math.floor(bits.length / 33) * 32;
    const entropyBits = bits.slice(0, dividerIndex);
    const checksumBits = bits.slice(dividerIndex);
    // calculate the checksum and compare
    const entropyBytes = entropyBits.match(/(.{1,8})/g).map(binaryToByte);
    if (entropyBytes.length < 16) {
        throw new Error(INVALID_ENTROPY);
    }
    if (entropyBytes.length > 32) {
        throw new Error(INVALID_ENTROPY);
    }
    if (entropyBytes.length % 4 !== 0) {
        throw new Error(INVALID_ENTROPY);
    }
    const entropy = Buffer.from(entropyBytes);
    const newChecksum = await deriveChecksumBitsAsync(entropy);
    if (newChecksum !== checksumBits) {
        console.log(entropy.toString('hex'));
        console.log('Expected ' + checksumBits + ' actual: ' + newChecksum);
        throw new Error(INVALID_CHECKSUM);
    }
    return entropy.toString('hex');
}
exports.mnemonicToEntropyAsync = mnemonicToEntropyAsync;
async function entropyToMnemonicAsync(entropy, wordlist) {
    if (!Buffer.isBuffer(entropy)) {
        entropy = Buffer.from(entropy, 'hex');
    }
    wordlist = wordlist || DEFAULT_WORDLIST;
    if (!wordlist) {
        throw new Error(WORDLIST_REQUIRED);
    }
    // 128 <= ENT <= 256
    if (entropy.length < 16) {
        throw new TypeError(INVALID_ENTROPY);
    }
    if (entropy.length > 32) {
        throw new TypeError(INVALID_ENTROPY);
    }
    if (entropy.length % 4 !== 0) {
        throw new TypeError(INVALID_ENTROPY);
    }
    const entropyBits = bytesToBinary(Array.from(entropy));
    const checksumBits = await deriveChecksumBitsAsync(entropy);
    const bits = entropyBits + checksumBits;
    const chunks = bits.match(/(.{1,11})/g);
    const words = chunks.map((binary) => {
        const index = binaryToByte(binary);
        return wordlist[index];
    });
    return wordlist[0] === '\u3042\u3044\u3053\u304f\u3057\u3093' // Japanese wordlist
        ? words.join('\u3000')
        : words.join(' ');
}
exports.entropyToMnemonicAsync = entropyToMnemonicAsync;
async function generateMnemonicAsync(strength, rng, wordlist) {
    strength = strength || 128;
    if (strength % 32 !== 0) {
        throw new TypeError(INVALID_ENTROPY);
    }
    if (!rng) {
        rng = (size) => {
            const rngValues = new Uint8Array(size);
            webcrypto.getRandomValues(rngValues);
            return Buffer.from(rngValues);
        };
    }
    return entropyToMnemonicAsync(rng(strength / 8), wordlist);
}
exports.generateMnemonicAsync = generateMnemonicAsync;
async function validateMnemonicAsync(mnemonic, wordlist) {
    try {
        await mnemonicToEntropyAsync(mnemonic, wordlist);
    }
    catch (e) {
        return false;
    }
    return true;
}
exports.validateMnemonicAsync = validateMnemonicAsync;
function setDefaultWordlist(language) {
    const result = _wordlists_1.wordlists[language];
    if (result) {
        DEFAULT_WORDLIST = result;
    }
    else {
        throw new Error('Could not find wordlist for language "' + language + '"');
    }
}
exports.setDefaultWordlist = setDefaultWordlist;
function getDefaultWordlist() {
    if (!DEFAULT_WORDLIST) {
        throw new Error('No Default Wordlist set');
    }
    return Object.keys(_wordlists_1.wordlists).filter((lang) => {
        if (lang === 'JA' || lang === 'EN') {
            return false;
        }
        return _wordlists_1.wordlists[lang].every((word, index) => word === DEFAULT_WORDLIST[index]);
    })[0];
}
exports.getDefaultWordlist = getDefaultWordlist;
var _wordlists_2 = require("./_wordlists");
exports.wordlists = _wordlists_2.wordlists;
